servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcClientBuilder.java

```diff
@@ -168,4 +227,91 @@ public void doAppendHttpClientFilter(final Predicate<StreamingHttpRequest> predi
+             */
+            @Nullable
+            public Duration apply(HttpRequestMetaData request) {
+                Duration duration = readTimeoutHeader(request);
+
+                if (null != duration) {
+                    // use timeout from request
+                    return duration;
+                }
+
+                Instant deadline = AsyncContext.get(GRPC_DEADLINE_KEY);
+
+                if (null != deadline) {
+                    // Use remaining time until deadline from async context
+                    return Duration.between(Instant.now(), deadline);
+                }
+
+                return defaultTimeout;
+            }
+
+            @Override
+            public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+                return strategy;
+            }
+        };
+    }
+
+    /**
+     * Extract the timeout duration from the HTTP headers if present.
+     *
+     * @param request The HTTP request to be used as source of the GRPC timeout header
+     * @return The non-negative timeout duration which may null if not present or malformed
+     */
+    @Nullable
+    static Duration readTimeoutHeader(HttpRequestMetaData request) {
+        return parseTimeoutHeader(request.headers().get(GRPC_TIMEOUT_HEADER_KEY));
+    }
+
+    /**
+     * Parse a gRPC HTTP timeout header value as a duration if present.
+     *
+     * @param value the text value of {@link #GRPC_TIMEOUT_HEADER_KEY} header to be parsed to a timeout duration or
+     * null if header was present
+     * @return The non-negative timeout duration which may null if not present or malformed
+     */
+    static @Nullable Duration parseTimeoutHeader(@Nullable CharSequence value) {
+        if (null != value && value.length() >= 2 && value.length() <= 9) {
+            char unitChar = value.charAt(value.length() - 1);
+            int unitsIdx = 0;
+            while (unitsIdx < TIMEOUT_UNIT_CHARS.length && unitChar != TIMEOUT_UNIT_CHARS[unitsIdx]) {
+                unitsIdx++;
+            }
+            if (unitsIdx == TIMEOUT_UNIT_CHARS.length) {
+                // Unrecognized units or malformed header
+                assert false : "Bad unit " + unitChar;
+                return null;
+            }
+
+            // parse number
+            long runningTotal = 0;
+            for (int digitIdx = 0; digitIdx < value.length() - 1; digitIdx++) {
+                char digitChar = value.charAt(digitIdx);
+                if (digitChar < '0' || digitChar > '9') {
+                    // Bad digit
+                    assert false : "Bad digit: " + digitChar;
+                    return null;
+                } else {
+                    runningTotal = runningTotal * 10L + (long) (digitChar - '0');
+                }
+            }
+
+            return Duration.of(runningTotal, TIMEOUT_CHRONOUNITS[unitsIdx]);
+        } else {
+            assert null == value : "bad timeout value: " + value;
```

Comment:
Why this assertion is required? It's legit to have no `grpc-timeout` on the request if users passed a value higher then the max value.

Reply:
This assert is associated with the case that the input is less than 2 characters or more than 9 characters long. The assert would not trigger if there was no Grpc-timeout header which it accomplishes, confusingly, but asserting `null == value`.

Reply:
now replaced by IllegalArgumentException